chainer/functions/array/spatial_transformer_sampler.py

```diff
@@ -0,0 +1,289 @@
+
+            handle = cudnn.get_handle()
+            x_desc = cudnn.create_tensor_descriptor(x)
+            dx_desc = cudnn.create_tensor_descriptor(gx)
+            dy_desc = cudnn.create_tensor_descriptor(gy)
+
+            one = numpy.array(1, dtype=x.dtype).ctypes
+            zero = numpy.array(0, dtype=x.dtype).ctypes
+            libcudnn.spatialTfSamplerBackward(
+                handle, self.st_desc.value,
+                one.data,
+                x_desc.value, x.data.ptr,
+                zero.data,
+                dx_desc.value, gx.data.ptr,
+                one.data,
+                dy_desc.value, gy.data.ptr,
+                grid_t.data.ptr, zero.data, ggrid_t.data.ptr)
+            ggrid = cuda.cupy.transpose(ggrid_t, axes=(0, 3, 1, 2))
+            return gx, ggrid
+        return self._backward(inputs, grad_outputs)
+
+    def _backward(self, inputs, grad_outputs):
+        x, grid = inputs
+        xp = cuda.get_array_module(x)
+        gy, = grad_outputs
+
+        B, C, H, W = x.shape
+        _, _, out_H, out_W = grid.shape
+        grid = grid.reshape(grid.shape[:2] + (-1,))
+
+        u = grid[:, 0]
+        v = grid[:, 1]
+
+        # clip coordinates to [-1, 1]
+        u = u.clip(-1, 1)
+        v = v.clip(-1, 1)
+
+        # rescale coordiantes from [-1, 1] to [0, width or height - 1]
+        u = (u + 1) / 2. * (W - 1)
+        v = (v + 1) / 2. * (H - 1)
+
+        # indices of the 2x2 pixel neighborhood surrounding the coordinates
+        u0 = xp.floor(u).astype(numpy.int32)
+        u0 = u0.clip(0, W - 2)
+        u1 = u0 + 1
+        v0 = xp.floor(v).astype(numpy.int32)
+        v0 = v0.clip(0, H - 2)
+        v1 = v0 + 1
+
+        # weights
+        wu0 = u - u0
+        wu1 = u1 - u
+        wv0 = v - v0
+        wv1 = v1 - v
+        wu0 = wu0.astype(gy.dtype)
+        wu1 = wu1.astype(gy.dtype)
+        wv0 = wv0.astype(gy.dtype)
+        wv1 = wv1.astype(gy.dtype)
+
+        # --- gu, gv
+        gus = []
+        gvs = []
+        for b in range(B):
+            gu_elem = (-wv1[b, :, None] * x[b, :, v0[b], u0[b]] +
+                       wv1[b, :, None] * x[b, :, v0[b], u1[b]] -
+                       wv0[b, :, None] * x[b, :, v1[b], u0[b]] +
+                       wv0[b, :, None] * x[b, :, v1[b], u1[b]])
+            gu_elem = gu_elem.reshape(out_H, out_W, C)
+            gu_elem = gu_elem.transpose(2, 0, 1)
+            gus.append(gu_elem)
+            gv_elem = (-wu1[b, :, None] * x[b, :, v0[b], u0[b]] -
+                       wu0[b, :, None] * x[b, :, v0[b], u1[b]] +
+                       wu1[b, :, None] * x[b, :, v1[b], u0[b]] +
+                       wu0[b, :, None] * x[b, :, v1[b], u1[b]])
+            gv_elem = gv_elem.reshape(out_H, out_W, C)
+            gv_elem = gv_elem.transpose(2, 0, 1)
+            gvs.append(gv_elem)
+        gu = xp.concatenate([xp.expand_dims(g, axis=0) for g in gus], axis=0)
+        gv = xp.concatenate([xp.expand_dims(g, axis=0) for g in gvs], axis=0)
```

Comment:
See the comment on _forward_gpu. The same rephrasing might be applied to this for-concat.